File: backend/SaidaPessoas.API/Controllers/NotificacoesController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using System.Security.Claims;
using SaidaPessoas.API.Data;
using SaidaPessoas.API.DTOs;

namespace SaidaPessoas.API.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class NotificacoesController : ControllerBase
{
    private readonly AppDbContext _context;

    public NotificacoesController(AppDbContext context) => _context = context;

    private int GetUserId() => int.Parse(User.FindFirst(ClaimTypes.NameIdentifier)!.Value);

    [HttpGet]
    public async Task<IActionResult> Listar([FromQuery] int limit = 20)
    {
        var userId = GetUserId();
        var lista = await _context.Notificacoes
            .Where(n => n.UsuarioId == userId)
            .OrderByDescending(n => n.DataCriacao)
            .Take(limit)
            .Select(n => new NotificacaoDto(
                n.Id, n.Mensagem, n.Tipo.ToString(), n.Lida, n.DataCriacao, n.SolicitacaoId))
            .ToListAsync();
        return Ok(lista);
    }

    [HttpGet("nao-lidas")]
    public async Task<IActionResult> ContarNaoLidas()
    {
        var userId = GetUserId();
        var count = await _context.Notificacoes.CountAsync(n => n.UsuarioId == userId && !n.Lida);
        return Ok(new { count });
    }

    [HttpPut("{id}/lida")]
    public async Task<IActionResult> MarcarLida(int id)
    {
        var userId = GetUserId();
        var n = await _context.Notificacoes.FirstOrDefaultAsync(x => x.Id == id && x.UsuarioId == userId);
        if (n == null) return NotFound();
        n.Lida = true;
        await _context.SaveChangesAsync();
        return Ok();
    }

    [HttpPut("marcar-todas-lidas")]
    public async Task<IActionResult> MarcarTodasLidas()
    {
        var userId = GetUserId();
        var naoLidas = await _context.Notificacoes
            .Where(n => n.UsuarioId == userId && !n.Lida)
            .ToListAsync();
        foreach (var n in naoLidas) n.Lida = true;
        await _context.SaveChangesAsync();
        return Ok(new { atualizadas = naoLidas.Count });
    }
}

File: backend-node/src/notificacoes/notificacoes.controller.ts
import { Controller, Get, NotFoundException, Param, ParseIntPipe, Put, Query } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { AuthUser, CurrentUser } from '../common/decorators';
import { toInstant } from '../common/dates';

@Controller('notificacoes')
export class NotificacoesController {
  constructor(private readonly prisma: PrismaService) {}

  @Get()
  async listar(@CurrentUser() user: AuthUser, @Query('limit') limit?: string) {
    const take = Math.min(Math.max(Number(limit) || 20, 1), 100);
    const lista = await this.prisma.notificacoes.findMany({
      where: { UsuarioId: user.userId },
      orderBy: { DataCriacao: 'desc' },
      take,
    });
    return lista.map((n) => ({
      id: n.Id,
      mensagem: n.Mensagem,
      tipo: n.Tipo,
      lida: n.Lida,
      dataCriacao: toInstant(n.DataCriacao),
      solicitacaoId: n.SolicitacaoId ?? undefined,
    }));
  }

  @Get('nao-lidas')
  async contarNaoLidas(@CurrentUser() user: AuthUser) {
    const count = await this.prisma.notificacoes.count({
      where: { UsuarioId: user.userId, Lida: false },
    });
    return { count };
  }

  @Put(':id/lida')
  async marcarLida(@CurrentUser() user: AuthUser, @Param('id', ParseIntPipe) id: number) {
    const n = await this.prisma.notificacoes.findFirst({ where: { Id: id, UsuarioId: user.userId } });
    if (!n) throw new NotFoundException();
    await this.prisma.notificacoes.update({ where: { Id: id }, data: { Lida: true } });
    return {};
  }

  @Put('marcar-todas-lidas')
  async marcarTodasLidas(@CurrentUser() user: AuthUser) {
    const r = await this.prisma.notificacoes.updateMany({
      where: { UsuarioId: user.userId, Lida: false },
      data: { Lida: true },
    });
    return { atualizadas: r.count };
  }
}
